### DISCUSSION TITLE
TCP/IP에 대해 설명해 주세요

### DISCUSSION THREAD

**Discussion by jswith** [woowacourse-study/2022-woowahan-bansanghwe#66] [Network]
## TCP/IP가 뭐지?

- 데이터를 디지털 신호로 바꾼다 -> 보낼 곳에 전달한다 -> 디지털 신호를 데이터로 되돌린다
- 위와 같은 작업을 효율적으로 수행하기 위해서 TCP/IP에서는 데이터를 신호로 바꾸거나, 신호를 데이터를 되돌리는 데 5단계의 절차를 거침.
- 각 단계를 계층이라고 하며, 위에서부터 순서대로 
- **애플리케이션 계층**
- **트랜스포트 계층**
- **네트워크 계층**
- **데이터 링크 계층**
- **물리 계층**
- 으로 구성되어 있음.

## 작게 나눠서 보내기

- TCP/IP의 특징 중 하나는 데이터를 일정한 크기로 분할해서 보낸다는 것.
- 작게 나눠진 데이터 하나 하나를 패킷이라고 하며, 이런 통신 방법을 패킷 통신이라고 함.
- 패킷 통신에서는 데이터를 세분화함으로써 하나의 회선을 사용하여 여러 데이터를 거의 동시에 송수신할 수 있음.
- 또한, 통신 중 데이터의 일부가 손상되어도 해당하는 부분만 다시 보내면 된다는 장점도 있음.

## 통신 프로토콜

- 컴퓨터마다 기종이나 통신 방식이 달라서 데이터 주고 받기가 너무 어려웡 ㅠㅠ
- 컴퓨터 간에 데이터를 주고받기 위한 공통된 구조를 만들면 각각의 차이에 관계없이 데이터를 주고받을 수 있음.
- 데이터를 주고받으려면 송신측과 수신측의 컴퓨터가 미리 정해놓은 공통된 메뉴얼에 따라 연락을 취해야 하는데 이 매뉴얼을 **프로토콜**이라고 함.

## TCP/IP

- 전 세계적으로 공통된 통신 프로토콜이 있다면? 그 프로토콜만 사용하면 어떤 컴퓨터끼리든 데이터를 주고받을 수 있음.
- 현재 전 세계 공통 통신 프로토콜로 채택된 것이 TCP/IP.
- 송신자 : 송신하고 싶은 데이터를 보냄 ->  공통된 형태가 됨. 공통된 형태이므로 받을 수 있음. -> 수신자 : 수신 측에서 처리할 수 있는 형태가 됨.
- 데이터를 주고받는 데는 여러 가지 작업이 필요한데, 하나의 프로토콜로 모든 것을 처리하기는 힘들기 때문에 TCP/IP는 여러 개의 프로토콜로 이루어져 있음.

## 계층화

- TCP/IP에서는 송수신과 관련된 일련의 작업을 몇 개의 단계로 나눠서 수행.
- 각 단계를 계층이라고 하며, 층으로 나눈 것을 계층화라고 함.
- 계층마다 작업을 독립시킨다는 장점이 있음. 다른 계층의 일을 몰라도 처리 가능.

## TCP/IP는 5계층

- 송신자
  - 애플리케이션 계층 : 수신측의 애플리케이션에서 처리할 수 있도록 만듦.
  - 트랜스포트 계층 : 네트워크 상에서의 공통된 형태로 만듦.
  - 네트워크 계층 : 수신처로 가는 경로를 정하고 보낼 수 있는 형태로 만듦.
  - 데이터 링크 계층 : 데이터를 비트열(0과 1로 표현한 것)로 변환.
  - 물리 계층 : 비트열을 전압의 변화나 빛의 점멸 신호로 변환하여 송신.
- 수신자
  - 물리 계층 : 전압의 변화나 빛의 점멸 신호를 비트열로 변환.
  - 데이터 링크 계층 : 비트열을 데이터로 변환.
  - 네트워크 계층 : 데이터의 수신인을 확인하고 자기 앞으로 온 것이 아니면 파기.
  - 트랜스포트 계층 : 데이터에 문제가 있으면 재전송을 요청.
  - 애플리케이션 계층 : 애플리케이션에서 표시, 재생.

## 각 층의 역할과 프로토콜

- TCP/IP의 5계층은 위로 갈수록 사용자와 가까운 작업을, 아래로 갈수록 기계와 가까운 작업을 담당.
- 애플리케이션 계층
  - 애플리케이션에 맞춰 통신을 수행할 수 있도록 함.
  - 애플리케이션마다 다양한 프로토콜이 있음.
  - HTTP, SMTP, POP3, FTP, TELNET, NNTP, RCP, ...
- 트랜스포트 계층
  - 송신된 데이터를 수신측 애플리케이션에 확실히 전달하기 위해 작동.
  - TCP, UDP
- 네트워크 계층
  - 수신측 컴퓨터까지 데이터를 전달하기 위해 작동.
  - 전달된 데이터가 손상됐는지 또는 수신측이 잘 받았는지에는 관여하지 않음.
  - IP
- 데이터 링크 계층
  - 네트워크에 직접 연결된 기기 간을 전송할 수 있도록 만듦.
  - 네트워크 계층과 물리 계층 간의 차이를 완전히 흡수하기 위한 다양한 프로토콜이 있음.
  - Ethernet, FDDI, ATM, PPP, PPPIoE, ...
- 물리 계층
  - 데이터를 신호로, 신호를 데이터로 변환.
  - 변환 방법은 통신 매체에 의존하기 때문에 특정 프로토콜이 정해져 있지 않음.

## 프로토콜 조합하기

- 프로토콜 조합을 바꿈으로써 다양한 애플리케이션이나 기기를 처리할 수 있음.
- 예를 들어 전자메일의 송신이라면 SMTP + TCP + IP, 수신이라면 POP3 + TCP + IP.

## 필요한 정보 추가하기

- 송신측의 각 계층에서는 수신측의 동일한 계층에서 필요한 정보를 공통된 서식으로 데이터에 추가해 감.
- 데이터보다 앞에 추가한 정보를 헤더, 뒤에 추가한 정보를 트레일러라고 함.
- 트레일러가 없는 경우도 있음.
- 상위 계층에서 추가된 헤더와 트레일러는 데이터와 합쳐져 하나로 묶임.
- 데이터와 정보를 하나로 묶어서 다루는 것을 캡슐화라고 함.

## 헤더와 계층화의 관계

![image](https://user-images.githubusercontent.com/64275588/175191689-462ba424-62db-4cf0-9b2f-fee2021d75d8.png)
![image](https://user-images.githubusercontent.com/64275588/175191696-81fba6a5-09a0-4816-8300-ef17b407b3d2.png)

- 송신측의 각 계층에서 추가된 헤더와 트레일러는 수신측의 동일한 계층에서만 사용됨.
  - 예) 송신측의 애플리케이션 계층에서 추가된 정보는 수신측의 애플리케이션 계층에서만 사용함.
- 이 때문에 서로 동일한 계층끼리만 연락을 취하고 있는 것처럼 보임.
- 상위 계층에서 추가된 헤더는 캡슐화되어 있으므로 다른 계층에서는 볼 수 없음.
- 사용된 헤더(및 트레일러)는 순서대로 벗겨짐.

## 송신측 작업

- 애플리케이션 계층
  - 애플리케이션 간에 데이터를 주고받기 위해 필요한 정보를 써넣음.
- 트랜스포트 계층
  - 애플리케이션을 나타내는 번호와 데이터를 조합하기 위해 필요한 정보를 써넣음.
  - 데이터를 접속 환경에 맞춘 크기로 분할. 분할된 하나하나를 패킷이라고 함.
- 네트워크 계층
  - 송수신할 컴퓨터의 주소와 수신 불명인 경우 데이터를 파기하는 표시 등을 써넣음.
- 데이터 링크 계층
  - 네트워크의 종류에 맞춘 형식으로 보낼 곳의 정보 등을 써넣음.
- 물리 계층
  - 비트열을 신호로 변환.

## 수신측 작업

- 물리 계층
  - 신호를 비트열로 변환.
- 데이터 링크 계층
  - 헤더에 적힌 정보를 확인하고 지정된 프로토콜에게 전달.
- 네트워크 계층
  - 헤더에 적힌 목적지가 맞는지 확인하고 지정된 프로토콜에게 전달.
- 트랜스포트 계층
  - 헤더를 확인하고 데이터를 순서대로 나열해서 조합.
  - 데이터에 문제가 있으면 그에 따른 처리를 함.
- 애플리케이션 계층
  - 지정된 애플리케이션에게 전달.

## 패킷 통신

- TCP/IP에서는 데이터를 일정한 크기(패킷)로 분할해서 송수신하는 패킷 교환이라는 방법으로 주고받기를 수행.
- 이러한 통신 방법을 패킷 통신이라고 함.
- 하나의 회선으로 여러 개의 데이터를 거의 동시에 주고받을 수 있음.
- 송신측
  - 데이터를 패킷으로 나눔.
  - 각 패킷에 목적지나 원래대로 되돌리기 위한 번호 등 필요한 정보를 추가.
  - 회선을 사용하여 데이터를 송신.
- 중계 지점
  - 다른 네트워크로 가는 중계 지점.
  - 다음은 어디로 가면 좋을지 등 새로운 정보가 추가됨.
- 수신측
  - 자기 앞으로 온 패킷만 받음.
  - 손상되어 있으면 해당 패킷만 다시 받음.
  - 수신 불명인 경우에는 파기하고 송신측에 연락.
  - 헤더 정보를 기초로 해서 데이터를 되돌림.

**Comment by al-bur** [comment]
- TCP/IP는 무엇인가요?
- TCP/IP가 왜 등장했나요?
- TCP/IP의 단점이나 한계가 있나요?
- 패킷에 대해 정확히 설명해주세요
- 패킷이 왜 등장했나요?
- 패킷을 나누는 기준이 있나요?
- OSI 7 layer vs TCP/IP의 차이?
- 왜 TCP/IP를 더 많이 사용할까요?
- UDP와 TCP를 설명해주세요?
- UDP의 예시를 알려주세요?
- 클로저 현상에 대해서 설명해주세요?
- 왜 클로저가 나오게 됐나요?
- 클로저 예시를 설명해주세요?
- 데이터 타입 두 가지에 대해주세요
- TCP/IP 준비해온 것 중에 이야기 못한 것 있나요?

**Comment by jswith** [comment]
TCP/IP는 무엇인가요?

**Comment by jswith** [comment]
TCP/IP가 왜 등장했나요?

**Comment by jswith** [comment]
TCP/IP의 단점이나 한계가 있나요?

**Comment by jswith** [comment]
패킷에 대해 정확히 설명해주세요

**Comment by jswith** [comment]
패킷이 왜 등장했나요?

**Comment by jswith** [comment]
패킷을 나누는 기준이 있나요?

**Comment by jswith** [comment]
OSI 7 layer vs TCP/IP의 차이?

**Comment by jswith** [comment]
왜 TCP/IP를 더 많이 사용할까요?

**Comment by jswith** [comment]
UDP와 TCP를 설명해주세요?

**Comment by jswith** [comment]
UDP의 예시를 알려주세요?

**Comment by jswith** [comment]
클로저 현상에 대해서 설명해주세요?

**Comment by jswith** [comment]
왜 클로저가 나오게 됐나요?

**Comment by jswith** [comment]
클로저 예시를 설명해주세요?

**Comment by jswith** [comment]
데이터 타입 두 가지에 대해 설명해주세요

**Comment by jswith** [comment]
TCP/IP 준비해온 것 중에 이야기 못한 것 있나요?
